### ISSUE TITLE
Changed behavior of the balances.Deposit events with kusama runtime 9122

### ISSUE THREAD

**Issue by TheGoldenEye** [paritytech/polkadot#4230]
In my tool (https://github.com/TheGoldenEye/polka-store) so far I've calculated the fee paid by the extrinsic sender (signer) from _balances.Deposit_ and _treasury.Deposit_ events (feePaid=_balances.Deposit+treasury.Deposit_)
This was working good until kusama runtine 9111. See example https://kusama.subscan.io/event?extrinsic=9866349-4
Starting with runtime 9122 there are now much more _balances.Deposit_ events. Seems (for _payout_stakers_ calls) each _staking.Rewarded_ event has a corresponding _balances.Deposit_ event now. So my fee calulation breaks. 
See example https://kusama.subscan.io/event?extrinsic=9874887-2. My calculated fee is 3.75 KSM (same is shown in subscan as fee for Extrinsic#9874887-2), but this cannot be right)
I saw the new balances.withdraw event, how to interpret this? Maybe this is (part of) the fee (paid by the sender) I'm looking for?

**Comment by apopiak** [comment]
This was introduced by https://github.com/paritytech/substrate/pull/9425 (duplicate events will be addressed by https://github.com/paritytech/polkadot/pull/4180)
Withdraw events represent fund movement out of an account (e.g. for fee payment). So a transaction will have a `balances.Withdraw` for the fee payment as well as `balances.Deposit` for the treasury and block author.

**Comment by bkchr** [comment]
So, can this be closed @TheGoldenEye ?

**Comment by freehere107** [comment]
@apopiak @bkchr   https://github.com/paritytech/substrate/pull/9425 Can I replace it with another event? After this pr merge, I cannot simply use ``treasury.deposit + balances.deposit`` to calculate the handling fee
https://polkadot.subscan.io/extrinsic/0x4832d65fb36f0dd2be9cfdc2ca31beda2f81ee69503a7f3acbaffb2b576fd3a0?event=7563119-17

**Comment by TheGoldenEye** [comment]
I've found the following solution:
I now use the last balances.Deposit (goes to block creator) and treasury.Deposit for fee calculation.
With #4180 the duplicate balances.Deposit (red) should disappear, but what about the (in my opinion unnecessary) events (blue in the picture)?
![image](https://user-images.githubusercontent.com/17945384/140884532-795cfdf9-64f3-4acd-8b59-61b0a7d8fc9c.png)

**Comment by TheGoldenEye** [comment]
btw, additional I check the accountId of balances.Deposit event, should be the id of block author

**Comment by apopiak** [comment]
So to recap, right now, there are:
+ 2 * `balances.Deposit` for the block author and 2 * `balances.Deposit` for the treasury (both of which will be reduced to 1 after #4180)
+ 1 `treasury.Deposit` for the treasury (which is an event generated in the treasury pallet and is independent of balances)

To get the full fees of a block I would suggest you:
+ deduplicate the `balances.Deposit` Events
+ for each transaction: add the two `balances.Deposit` events going to the block author and the treasury respectively

You should be able to match this amount against the corresponding amount withdrawn from the account sending the transaction (there is a corresponding `balances.Withdraw` now).

**Comment by apopiak** [comment]
> I've found the following solution: I now use the last balances.Deposit (goes to block creator) and treasury.Deposit for fee calculation. 

This will only work for one transaction, FYI.
> With #4180 the duplicate balances.Deposit (red) should disappear, but what about the (in my opinion unnecessary) events (blue in the picture)?

The blue events are create by the balance movements induced by the staking rewards.
With the new events, all balance movements have a corresponding balances event.

**Comment by TheGoldenEye** [comment]
> * 2 * `balances.Deposit` for the block author and 2 * `balances.Deposit` for the treasury (both of which will be reduced to 1 after [remove duplicate Deposit from OnUnbalanced implementation #4180](https://github.com/paritytech/polkadot/pull/4180))
> * 1 `treasury.Deposit` for the treasury (which is an event generated in the treasury pallet and is independent of balances)

only `balances.Deposit` is duplicated, not the one for treasury


> > I've found the following solution: I now use the last balances.Deposit (goes to block creator) and treasury.Deposit for fee calculation.
> 
> This will only work for one transaction, FYI.

we consider the events of one transaction, so no problem

I have picked out all the relevant events for Extrinsic 9874887-2
![image](https://user-images.githubusercontent.com/17945384/140971465-8ec4eb3a-c1ec-4c93-824f-511d03fe2800.png)
Therew are 2 ways for fee calculation I see:
1) The fee paid by the sender is `9874887-173 + 9874887-174 = 115666534`
2) This is same as `9874887-12 - 9874887-171 = 115666534`

**What is the reason for event 9874887-171? Without this the the amount of event 9874887-12 should be 115666534. (the fee paid)** 
Btw, both variants of the fee calculation are prone to errors. If there is an additional balance.Deposit (e.g. staking payout) for the signer account (2) or the block author account (1), then the calculation goes wrong.

**Comment by apopiak** [comment]
> [...] There are 2 ways for fee calculation I see:
> 
>     1. The fee paid by the sender is `9874887-173 + 9874887-174 = 115666534`
> 
>     2. This is same as `9874887-12 - 9874887-171 = 115666534`

Both of those are correct and should give the same result.
1. the first is the combination of the treasury and the block author payout.
2. is the initial fee minus the refund

> **What is the reason for event 9874887-171? 

It is a refund to the sender of the transaction because the final fee is lower than the initial fee.

> Btw, both variants of the fee calculation are prone to errors. If there is an additional balance.Deposit (e.g. staking payout) for the signer account (2) or the block author account (1), then the calculation goes wrong.

Yes, filtering might get complicated in edge cases; should be possible, though.
Feel free to open a new issue for this. I don't think it's very likely more events will be added, though.

My sense is that we've covered the changed behavior. Good to close?

**Comment by TheGoldenEye** [comment]
Yep, good to close. Thx @apopiak
